clang/TranslationUnit.d

```diff
@@ -245,17 +255,104 @@ struct TranslationUnit
         return relativeLocationAccessorImpl(cursor.all);
     }
 
-    TokenRange tokenize(SourceRange extent)
+    private struct TokenRange
+    {
+        CXTranslationUnit cx;
+        CXToken* tokens;
+        uint numTokens;
+        uint currentToken;
+
+        Token makeToken(CXToken token)
+        {
+            return Token(
+                clang_getTokenKind(token).toD,
+                clang_getTokenSpelling(cx, token).toD,
+                SourceRange(clang_getTokenExtent(cx, token)));
+        }
+
+        Token front()
+        {
+            return makeToken(tokens[currentToken]);
+        }
+
+        bool empty()
+        {
+            return numTokens == 0 || numTokens == currentToken;
+        }
+
+        void popFront()
+        {
+            currentToken++;
+        }
+
+        void dispose()
+        {
+            clang_disposeTokens(cx, tokens, numTokens);
+        }
+    }
+
+    private static TokenRange tokenizeImpl(CXTranslationUnit cx, SourceRange extent)
     {
-        import std.algorithm.mutation : stripRight;
+        auto range = TokenRange(cx);
+        clang_tokenize(cx, extent.cx, &range.tokens, &range.numTokens);
+        return range;
+    }
 
-        CXToken* tokens = null;
-        uint numTokens = 0;
-        clang_tokenize(cx, extent.cx, &tokens, &numTokens);
-        auto result = TokenRange(cx, tokens, numTokens);
+    package static Token[] tokenize(CXTranslationUnit cx, SourceRange extent)
+    {
+        import std.array : array;
+        import std.algorithm : stripRight;
+
+        auto range = tokenizeImpl(cx, extent);
+        auto tokens = range.array;
+        range.dispose();
 
         // For some reason libclang returns some tokens out of cursors extent.cursor
-        return result.stripRight!(token => !intersects(extent, token.extent));
+        return tokens.stripRight!(token => !intersects(extent, token.extent));
```

Comment:
It's not strictly necessary. We don't need to over complicate things. I thought it would be possible with an input range.